src/provider/provider_ctl_stats_impl.h

```diff
@@ -0,0 +1,128 @@
+
+    size_t *arg_out = arg;
+    CTL_PROVIDER_TYPE *provider = (CTL_PROVIDER_TYPE *)ctx;
+    COMPILE_ERROR_ON(sizeof(provider->stats.allocated_memory) !=
+                     sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(*arg_out) != sizeof(uint64_t));
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.allocated_memory,
+                                  (uint64_t *)arg_out);
+    return 0;
+}
+
+static int CTL_RUNNABLE_HANDLER(reset)(void *ctx, umf_ctl_query_source_t source,
+                                       void *arg,
+                                       umf_ctl_index_utlist_t *indexes,
+                                       const char *extra_name,
+                                       umf_ctl_query_type_t query_type) {
+    /* suppress unused-parameter errors */
+    (void)source, (void)indexes, (void)arg, (void)extra_name, (void)query_type;
+
+    CTL_PROVIDER_TYPE *provider = (CTL_PROVIDER_TYPE *)ctx;
+    size_t allocated;
+
+    COMPILE_ERROR_ON(sizeof(provider->stats.allocated_memory) !=
+                     sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(allocated) != sizeof(uint64_t));
+
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.allocated_memory,
+                                  (uint64_t *)&allocated);
+    utils_atomic_store_release_u64((uint64_t *)&provider->stats.peak_memory,
+                                   (uint64_t)allocated);
+
+    return 0;
+}
+
+static const umf_ctl_node_t CTL_NODE(stats)[] = {
+    CTL_LEAF_RO(allocated_memory), CTL_LEAF_RO(peak_memory),
+    CTL_LEAF_RUNNABLE(reset), CTL_NODE_END};
+
+static inline void provider_ctl_stats_alloc(CTL_PROVIDER_TYPE *provider,
+                                            size_t size) {
+
+    COMPILE_ERROR_ON(sizeof(provider->stats.allocated_memory) !=
+                     sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(provider->stats.peak_memory) != sizeof(uint64_t));
+    COMPILE_ERROR_ON(sizeof(size) != sizeof(uint64_t));
+    // TODO: Change to memory_order_relaxed when we will have a proper wrapper
+    size_t allocated =
+        utils_fetch_and_add_u64((uint64_t *)&provider->stats.allocated_memory,
+                                (uint64_t)size) +
+        size;
+
+    uint64_t peak;
+    utils_atomic_load_acquire_u64((uint64_t *)&provider->stats.peak_memory,
+                                  &peak);
+
+    while (allocated > peak &&
+           !utils_compare_exchange_u64((uint64_t *)&provider->stats.peak_memory,
+                                       &peak, (uint64_t *)&allocated)) {
+        /* If the compare-exchange fails, 'peak' is updated to the current value of peak_memory.
+       We then re-check whether allocated is still greater than the updated peak value. */
+        ;
+    }
+}
+
+static inline void provider_ctl_stats_free(CTL_PROVIDER_TYPE *provider,
+                                           size_t size) {
+
+    COMPILE_ERROR_ON(sizeof(size) != sizeof(uint64_t));
```

Comment:
why we need this checks in all functions?

Reply:
Done